File: python/src/request_log.py
import json
import logging
import re
import time
from urllib.parse import urlsplit, urlunsplit

_CONTEXT_FIELDS = (
    "functionName", "functionRequestId", "functionInvocationId",
    "x-ms-client-request-id", "x-ms-correlation-id", "msCorrelationIdSource", "omittedIdFields",
    "channel", "evaluation", "providerName",
)
_CREDENTIAL_FIELDS = (
    "providerAuthMode", "providerCredentialSource", "providerTenantId",
    "functionOutboundClientId", "functionOutboundManagedIdentityClientId",
)
_HTTP_METHODS = {"GET", "HEAD", "POST", "PUT", "DELETE", "CONNECT", "OPTIONS", "TRACE", "PATCH"}


_IDENTIFIER_PATTERN = re.compile(r"[A-Za-z0-9][A-Za-z0-9._:-]{0,127}")


class RequestLog:
    """Request-scoped, explicitly selected metadata; never serialize delivery/provider models."""

    def __init__(self, request_id, invocation_id, ms_request_id, ms_correlation_id, function_name="send_otp"):
        self.started = time.monotonic()
        self.provider_started = None
        self.credential_started = None
        self.data = {
            "functionName": function_name,
            "functionRequestId": request_id,
            "functionInvocationId": invocation_id or None,
            "x-ms-client-request-id": None,
            "x-ms-correlation-id": None,
            "msCorrelationIdSource": "none",
            "omittedIdFields": [],
            "envelopeType": None,
            "ttlSeconds": None,
            "channel": None,
            "evaluation": None,
            "encryptionKeyIdMismatch": False,
            "providerName": None,
            "providerAuthMode": None,
            "providerCredentialSource": None,
            "providerCredentialElapsedMs": None,
            "providerTenantId": None,
            "functionOutboundClientId": None,
            "functionOutboundManagedIdentityClientId": None,
            "providerHttpMethod": None,
            "providerEndpoint": None,
            "providerAttempted": False,
            "providerHttpStatus": None,
            "providerStatus": None,
            "providerOutcome": None,
            "providerMessageId": None,
            "providerElapsedMs": None,
            "providerTimeoutMs": None,
            "failureStage": None,
            "failureReason": None,
            "responseContainsNonce": None,
            "responseContainsCorrelationId": None,
        }
        self._set_identifier("x-ms-client-request-id", ms_request_id)
        self._set_identifier("x-ms-correlation-id", ms_correlation_id)
        self.data["msCorrelationIdSource"] = "header" if self.data["x-ms-correlation-id"] else "none"

    def _set_identifier(self, field, value):
        valid = isinstance(value, str) and len(value) <= 128 and _IDENTIFIER_PATTERN.fullmatch(value) is not None
        self.data[field] = value if valid else None
        self.data["omittedIdFields"] = [name for name in self.data["omittedIdFields"] if name != field]
        if not valid and value is not None and not (isinstance(value, str) and not value.strip()):
            self.data["omittedIdFields"].append(field)

    def service(self, event_name, details=None, level=logging.INFO):
        record = {
            "logType": "service", "eventName": event_name,
            **{key: self.data[key] for key in _CONTEXT_FIELDS},
            **(details or {}),
            "elapsedMs": int((time.monotonic() - self.started) * 1000),
        }
        logging.log(level, "%s", json.dumps(record))

    def envelope_validated(self, envelope, correlation_id, source):
        self.data["envelopeType"] = envelope.type
        self.data["ttlSeconds"] = envelope.ttl_seconds
        self.data["channel"] = "sms" if envelope.channel == 1 else "voice"
        self.data["evaluation"] = envelope.mode == 2
        self._set_identifier("x-ms-correlation-id", correlation_id)
        self.data["msCorrelationIdSource"] = source if self.data["x-ms-correlation-id"] else "none"
        self.service("envelope_validated", {
            "envelopeType": self.data["envelopeType"],
            "ttlSeconds": self.data["ttlSeconds"],
            "encryptedDeliveryContextPresent": True,
        })

    def key_id_mismatch(self):
        self.data["encryptionKeyIdMismatch"] = True
        self.service("encryption_key_id_mismatch", level=logging.WARNING)

    def provider_selected(self, manifest):
        self.data["providerName"] = manifest["id"]
        mode = manifest["auth"].get("mode")
        self.data["providerAuthMode"] = mode if mode in ("apiKey", "oauth") else "unsupported"
        self.service("provider_selected", {"providerAuthMode": self.data["providerAuthMode"]})

    def credential_resolution_started(self, config):
        self.credential_started = time.monotonic()
        oauth = self.data["providerAuthMode"] == "oauth"
        self.data["providerCredentialSource"] = ("managed_identity_client_assertion" if oauth
            else "key_vault" if self.data["providerAuthMode"] == "apiKey" else "unsupported")
        if oauth:
            self._set_identifier("providerTenantId", config.provider_tenant_id)
            self._set_identifier("functionOutboundClientId", config.outbound_client_id)
            self._set_identifier("functionOutboundManagedIdentityClientId", config.outbound_managed_identity_client_id)
        self.service("provider_credential_resolution_started", self._credential_details())

    def _credential_details(self):
        return {key: self.data[key] for key in _CREDENTIAL_FIELDS}

    def _credential_resolution_finished(self):
        if self.credential_started is not None:
            self.data["providerCredentialElapsedMs"] = int((time.monotonic() - self.credential_started) * 1000)
            self.credential_started = None

    def credential_resolved(self):
        self._credential_resolution_finished()
        self.service("provider_credential_resolved", {
            **self._credential_details(),
            "providerCredentialElapsedMs": self.data["providerCredentialElapsedMs"],
        })

    def provider_request_built(self, method, endpoint):
        normalized = method.upper() if isinstance(method, str) else None
        self.data["providerHttpMethod"] = normalized if normalized in _HTTP_METHODS else "other"
        url = urlsplit(endpoint)
        host = f"[{url.hostname}]" if ":" in url.hostname else url.hostname
        authority = host if url.port in (None, 443) else f"{host}:{url.port}"
        self.data["providerEndpoint"] = urlunsplit((url.scheme, authority, url.path or "/", "", ""))
        self.service("provider_request_built", {
            "providerHttpMethod": self.data["providerHttpMethod"],
            "providerEndpoint": self.data["providerEndpoint"],
            "providerScheme": "https",
            "redirectsAllowed": False,
        })

    def provider_request_started(self, timeout_ms):
        self.provider_started = time.monotonic()
        self.data["providerAttempted"] = True
        self.data["providerTimeoutMs"] = timeout_ms
        self.service("provider_request_started", {
            "providerTimeoutMs": timeout_ms,
            "providerHttpMethod": self.data["providerHttpMethod"],
            "providerEndpoint": self.data["providerEndpoint"],
        })

    def provider_response_received(self, status):
        self.data["providerHttpStatus"] = status
        self.service("provider_response_received", {"providerHttpStatus": status})

    def provider_request_finished(self):
        if self.provider_started is not None:
            self.data["providerElapsedMs"] = int((time.monotonic() - self.provider_started) * 1000)
            self.provider_started = None

    def provider_response_processed(self, manifest, parsed, outcome, http_status, valid_json):
        status = parsed.provider_status_name or parsed.provider_status_code
        known = type(status) in (str, int) and status != "default" and status in manifest["response_mapping"]
        self.data["providerStatus"] = str(status) if known else "unmapped"
        self.data["providerOutcome"] = outcome
        self._set_identifier("providerMessageId", parsed.provider_message_id)
        if outcome != "Continue":
            self.data["failureStage"] = "provider_response"
            self.data["failureReason"] = "provider_rejected" if valid_json else "invalid_provider_json"
        self.service("provider_response_processed", {
            "providerHttpStatus": self.data["providerHttpStatus"],
            "providerStatus": self.data["providerStatus"],
            "providerOutcome": outcome,
            "providerMessageId": self.data["providerMessageId"],
            "providerElapsedMs": self.data["providerElapsedMs"],
            "httpStatus": http_status,
            "failureReason": self.data["failureReason"],
        }, logging.ERROR if http_status >= 500 else logging.INFO if http_status == 200 else logging.WARNING)

    def failure(self, stage, reason, http_status):
        self._credential_resolution_finished()
        self.provider_request_finished()
        self.data["failureStage"] = stage
        self.data["failureReason"] = reason
        self.service(f"{stage}_failed", {"failureReason": reason, "httpStatus": http_status},
                     logging.ERROR if http_status >= 500 else logging.WARNING)

    def response_prepared(self, http_status, contains_nonce, contains_correlation_id):
        self.data["responseContainsNonce"] = contains_nonce
        self.data["responseContainsCorrelationId"] = contains_correlation_id
        self.service("response_prepared", {
            "httpStatus": http_status,
            "responseContainsNonce": contains_nonce,
            "responseContainsCorrelationId": contains_correlation_id,
        })

    def complete(self, http_status):
        self._credential_resolution_finished()
        self.provider_request_finished()
        logging.info("%s", json.dumps({
            "logType": "request", "eventName": "request_completed",
            **self.data,
            "httpStatus": http_status,
            "result": ("evaluated" if self.data["evaluation"] else "accepted") if http_status == 200 else "failed",
            "elapsedMs": int((time.monotonic() - self.started) * 1000),
        }))

File: dotnet/Src/RequestLog.cs
using System.Diagnostics;
using System.Text.Json;
using System.Text.RegularExpressions;
using Microsoft.Extensions.Logging;

namespace Epp.Otp;

// Only explicitly selected metadata enters logs; never serialize delivery/provider models.
public sealed class RequestLog
{
    private static readonly string[] ContextFields =
    {
        "functionName", "functionRequestId", "functionInvocationId",
        "x-ms-client-request-id", "x-ms-correlation-id", "msCorrelationIdSource", "omittedIdFields",
        "channel", "evaluation", "providerName",
    };
    private static readonly string[] CredentialFields =
    {
        "providerAuthMode", "providerCredentialSource", "providerTenantId",
        "functionOutboundClientId", "functionOutboundManagedIdentityClientId",
    };
    private static readonly Regex IdentifierPattern = new(@"\A[A-Za-z0-9][A-Za-z0-9._:-]{0,127}\z", RegexOptions.CultureInvariant);

    private readonly ILogger _logger;
    private readonly Stopwatch _started = Stopwatch.StartNew();
    private readonly Dictionary<string, object?> _data;
    private readonly List<string> _omittedIdFields = new();
    private Stopwatch? _providerStarted;
    private Stopwatch? _credentialStarted;

    public bool HasFailure => _data["failureStage"] is not null;

    public RequestLog(ILogger logger, string requestId, string? invocationId, string? msRequestId, string? msCorrelationId)
    {
        _logger = logger;
        _data = new()
        {
            ["functionName"] = "SendOtp",
            ["functionRequestId"] = requestId,
            ["functionInvocationId"] = invocationId,
            ["x-ms-client-request-id"] = null,
            ["x-ms-correlation-id"] = null,
            ["msCorrelationIdSource"] = "none",
            ["omittedIdFields"] = Array.Empty<string>(),
            ["envelopeType"] = null,
            ["ttlSeconds"] = null,
            ["channel"] = null,
            ["evaluation"] = null,
            ["encryptionKeyIdMismatch"] = false,
            ["providerName"] = null,
            ["providerAuthMode"] = null,
            ["providerCredentialSource"] = null,
            ["providerCredentialElapsedMs"] = null,
            ["providerTenantId"] = null,
            ["functionOutboundClientId"] = null,
            ["functionOutboundManagedIdentityClientId"] = null,
            ["providerHttpMethod"] = null,
            ["providerEndpoint"] = null,
            ["providerAttempted"] = false,
            ["providerHttpStatus"] = null,
            ["providerStatus"] = null,
            ["providerOutcome"] = null,
            ["providerMessageId"] = null,
            ["providerElapsedMs"] = null,
            ["providerTimeoutMs"] = null,
            ["failureStage"] = null,
            ["failureReason"] = null,
            ["responseContainsNonce"] = null,
            ["responseContainsCorrelationId"] = null,
        };
        SetIdentifier("x-ms-client-request-id", msRequestId);
        SetIdentifier("x-ms-correlation-id", msCorrelationId);
        _data["msCorrelationIdSource"] = _data["x-ms-correlation-id"] is null ? "none" : "header";
    }

    private void SetIdentifier(string field, string? value)
    {
        var valid = value is { Length: <= 128 } && IdentifierPattern.IsMatch(value);
        _data[field] = valid ? value : null;
        _omittedIdFields.Remove(field);
        if (!valid && !string.IsNullOrWhiteSpace(value)) _omittedIdFields.Add(field);
        _data["omittedIdFields"] = _omittedIdFields.ToArray();
    }

    public void Service(string eventName, Dictionary<string, object?>? details = null, LogLevel level = LogLevel.Information)
    {
        var record = new Dictionary<string, object?> { ["logType"] = "service", ["eventName"] = eventName };
        foreach (var field in ContextFields) record[field] = _data[field];
        if (details is not null)
            foreach (var (key, value) in details) record[key] = value;
        record["elapsedMs"] = _started.ElapsedMilliseconds;
        Write(level, eventName, record);
    }

    public void EnvelopeValidated(Envelope envelope, string? correlationId, string source)
    {
        _data["envelopeType"] = envelope.Type;
        _data["ttlSeconds"] = envelope.TtlSeconds;
        _data["channel"] = EnvelopeParser.ChannelName(envelope.Channel);
        _data["evaluation"] = envelope.Mode == EnvelopeParser.ModeEvaluation;
        SetIdentifier("x-ms-correlation-id", correlationId);
        _data["msCorrelationIdSource"] = _data["x-ms-correlation-id"] is null ? "none" : source;
        Service("envelope_validated", new()
        {
            ["envelopeType"] = _data["envelopeType"],
            ["ttlSeconds"] = _data["ttlSeconds"],
            ["encryptedDeliveryContextPresent"] = true,
        });
    }

    public void KeyIdMismatch()
    {
        _data["encryptionKeyIdMismatch"] = true;
        Service("encryption_key_id_mismatch", level: LogLevel.Warning);
    }

    public void ProviderSelected(ProviderManifest manifest)
    {
        _data["providerName"] = manifest.Id;
        _data["providerAuthMode"] = manifest.Auth.Mode is "apiKey" or "oauth" ? manifest.Auth.Mode : "unsupported";
        Service("provider_selected", new() { ["providerAuthMode"] = _data["providerAuthMode"] });
    }

    public void CredentialResolutionStarted(AppConfig config)
    {
        _credentialStarted = Stopwatch.StartNew();
        _data["providerCredentialSource"] = _data["providerAuthMode"] switch
        {
            "oauth" => "managed_identity_client_assertion",
            "apiKey" => "key_vault",
            _ => "unsupported",
        };
        if (Equals(_data["providerAuthMode"], "oauth"))
        {
            SetIdentifier("providerTenantId", config.ProviderTenantId);
            SetIdentifier("functionOutboundClientId", config.OutboundClientId);
            SetIdentifier("functionOutboundManagedIdentityClientId", config.OutboundManagedIdentityClientId);
        }
        Service("provider_credential_resolution_started", CredentialDetails());
    }

    private Dictionary<string, object?> CredentialDetails() =>
        CredentialFields.ToDictionary(key => key, key => _data[key]);

    private void CredentialResolutionFinished()
    {
        if (_credentialStarted is null) return;
        _data["providerCredentialElapsedMs"] = _credentialStarted.ElapsedMilliseconds;
        _credentialStarted = null;
    }

    public void CredentialResolved()
    {
        CredentialResolutionFinished();
        var details = CredentialDetails();
        details["providerCredentialElapsedMs"] = _data["providerCredentialElapsedMs"];
        Service("provider_credential_resolved", details);
    }

    public void ProviderRequestBuilt(string? method, string endpoint)
    {
        var normalized = method?.ToUpperInvariant();
        _data["providerHttpMethod"] = normalized is "GET" or "HEAD" or "POST" or "PUT" or "DELETE"
            or "CONNECT" or "OPTIONS" or "TRACE" or "PATCH" ? normalized : "other";
        var uri = new Uri(endpoint, UriKind.Absolute);
        _data["providerEndpoint"] = uri.GetComponents(UriComponents.SchemeAndServer, UriFormat.UriEscaped) + uri.AbsolutePath;
        Service("provider_request_built", new()
        {
            ["providerHttpMethod"] = _data["providerHttpMethod"],
            ["providerEndpoint"] = _data["providerEndpoint"],
            ["providerScheme"] = "https",
            ["redirectsAllowed"] = false,
        });
    }

    public void ProviderRequestStarted(int timeoutMs)
    {
        _providerStarted = Stopwatch.StartNew();
        _data["providerAttempted"] = true;
        _data["providerTimeoutMs"] = timeoutMs;
        Service("provider_request_started", new()
        {
            ["providerTimeoutMs"] = timeoutMs,
            ["providerHttpMethod"] = _data["providerHttpMethod"],
            ["providerEndpoint"] = _data["providerEndpoint"],
        });
    }

    public void ProviderResponseReceived(int status)
    {
        _data["providerHttpStatus"] = status;
        Service("provider_response_received", new() { ["providerHttpStatus"] = status });
    }

    public void ProviderRequestFinished()
    {
        if (_providerStarted is null) return;
        _data["providerElapsedMs"] = _providerStarted.ElapsedMilliseconds;
        _providerStarted = null;
    }

    public void ProviderResponseProcessed(ProviderManifest manifest, ParsedResponse parsed, Outcome outcome, int httpStatus, bool validJson)
    {
        var status = parsed.ProviderStatusName ?? parsed.ProviderStatusCode;
        var known = status is not null && status != "default" && manifest.ResponseMapping.ContainsKey(status);
        _data["providerStatus"] = known ? status : "unmapped";
        _data["providerOutcome"] = outcome.ToString();
        SetIdentifier("providerMessageId", parsed.ProviderMessageId);
        if (outcome != Outcome.Continue)
        {
            _data["failureStage"] = "provider_response";
            _data["failureReason"] = validJson ? "provider_rejected" : "invalid_provider_json";
        }
        Service("provider_response_processed", new()
        {
            ["providerHttpStatus"] = _data["providerHttpStatus"],
            ["providerStatus"] = _data["providerStatus"],
            ["providerOutcome"] = _data["providerOutcome"],
            ["providerMessageId"] = _data["providerMessageId"],
            ["providerElapsedMs"] = _data["providerElapsedMs"],
            ["httpStatus"] = httpStatus,
            ["failureReason"] = _data["failureReason"],
        }, httpStatus >= 500 ? LogLevel.Error : httpStatus == 200 ? LogLevel.Information : LogLevel.Warning);
    }

    public void Failure(string stage, string reason, int httpStatus)
    {
        CredentialResolutionFinished();
        ProviderRequestFinished();
        _data["failureStage"] = stage;
        _data["failureReason"] = reason;
        Service(stage + "_failed", new() { ["failureReason"] = reason, ["httpStatus"] = httpStatus },
            httpStatus >= 500 ? LogLevel.Error : LogLevel.Warning);
    }

    public void ResponsePrepared(int httpStatus, bool containsNonce, bool containsCorrelationId)
    {
        _data["responseContainsNonce"] = containsNonce;
        _data["responseContainsCorrelationId"] = containsCorrelationId;
        Service("response_prepared", new()
        {
            ["httpStatus"] = httpStatus,
            ["responseContainsNonce"] = containsNonce,
            ["responseContainsCorrelationId"] = containsCorrelationId,
        });
    }

    public void Complete(int httpStatus)
    {
        CredentialResolutionFinished();
        ProviderRequestFinished();
        var record = new Dictionary<string, object?>(_data)
        {
            ["logType"] = "request",
            ["eventName"] = "request_completed",
            ["httpStatus"] = httpStatus,
            ["result"] = httpStatus == 200 ? (Equals(_data["evaluation"], true) ? "evaluated" : "accepted") : "failed",
            ["elapsedMs"] = _started.ElapsedMilliseconds,
        };
        Write(LogLevel.Information, "request_completed", record);
    }

    private void Write(LogLevel level, string eventName, Dictionary<string, object?> record) =>
        _logger.Log(level, new EventId(0, eventName), record, null,
            static (state, _) => JsonSerializer.Serialize(state));
}
